dfm_tools/observations.py

```diff
@@ -849,6 +871,72 @@ def rwsddl_ssh_retrieve_data(row, dir_output, time_min, time_max):
     return ds
 
 
+def gtsm_era5_cds_ssh_retrieve_data(row, dir_output, time_min, time_max):
+    """
+    Retrieve data from Climate Data Store
+    Can only retrieve entire files, subsetting for time and stations is done after download
+    The function checks if the files have already been downloaded to cache
+    """
+    
+    if time_min is None or time_max is None:
+        raise ValueError("cannot supply None for 'time_min' or 'time_max' to 'gtsm_era5_cds_retrieve_data()'")
+    
+    dir_cache = get_dir_testdata()
+
+    # Define timeframe of the request as lists of years and months
+    time_min_dt = datetime.strptime(time_min, '%Y-%m-%d')
+    time_max_dt = datetime.strptime(time_max, '%Y-%m-%d')
+    if time_max_dt.year == time_min_dt.year: # dates within one year
+        years = [time_max_dt.year]
+        months = [list(range(time_min_dt.month, time_max_dt.month+1))]
+    elif (time_max_dt.year - time_min_dt.year) == 1: # dates across the year boundary
+        years = [time_min_dt.year, time_max_dt.year]
+        months = [list(range(time_min_dt.month, 13)), list(range(1, time_max_dt.month))]
+    elif (time_max_dt.year - time_min_dt.year) > 1: # dates span multiple years
+        years = list(range(time_min_dt.year, time_max_dt.year+1))
+        months = [list(range(time_min_dt.month, 13))]
+        for yy in years[1:-1]:
+            months.append(list(range(1, 13)))
+        months.append(list(range(1, time_max_dt.month)))
+
+    # Retrieve data via an API request and extract archive - for files not found in the cache
+    for yy,year in enumerate(years):
+        for mm,month in enumerate(months[yy]):
+            filename = os.path.join(dir_cache, f'reanalysis_waterlevel_hourly_{year}_{month:02}_v2.nc')
+            if not os.path.isfile(filename):
+                print('... retrieving GTSM-ERA5 data for %i, month %i' % (year,month))    
+                tmp_zipfile = os.path.join(dir_cache, f'gtsm_era5_{year}_{month:02}.zip')
+
+                # Make connection with CDS via API (assumes URL and KEY already stored in .cdsapirc file)
+                c = cdsapi.Client() 
+                c.retrieve(
+                    'sis-water-level-change-timeseries-cmip6',
+                    {
+                        'variable': 'total_water_level',
+                        'experiment': 'reanalysis',
+                        'temporal_aggregation': 'hourly',
+                        'year': str(year),
+                        'month': [str(x).zfill(2) for x in months[yy]],
+                        'format': 'zip',
+                    }, 
+                    tmp_zipfile)
+            
+                with ZipFile(tmp_zipfile, 'r') as zip_ref:
+                    zip_ref.extractall(dir_cache)
+                os.remove(tmp_zipfile)
+
+    # open dataset
+    ds = xr.open_mfdataset(os.path.join(dir_cache, f'reanalysis_waterlevel_hourly_*_v2.nc'))
```

Comment:
- include some arguments in open_mfdataset() to make it faster and more robust (see our teams chat of last week)
- sonarcloud complains about you using an f-string without arguments, so you can remove the f here